python/lsst/daf/butler/core/logging.py

```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
         Can be called even if the key is not known to MDC.
         """
         cls._MDC.pop(key, None)
+
+
+class ButlerLogRecord(BaseModel):
+    """A model representing a `logging.LogRecord`.
+
+    A `~logging.LogRecord` always uses the current time in its record
+    when recreated and that makes it impossible to use it as a
+    serialization format. Instead have a local representation of a
+    `~logging.LogRecord` that matches Butler needs.
+    """
+
+    _log_format: ClassVar[str] = _LONG_LOG_FORMAT
+
+    name: str
+    asctime: datetime.datetime
+    message: str
+    levelno: int
+    levelname: str
+    filename: str
+    pathname: str
+    lineno: int
+    funcName: Optional[str]
+    process: int
+    processName: str
+    exc_info: Optional[str]
+    MDC: Dict[str, str]
+
+    class Config:
+        """Pydantic model configuration."""
+
+        allow_mutation = False
+
+    @classmethod
+    def from_record(cls, record: LogRecord) -> ButlerLogRecord:
+        """Create a new instance from a `~logging.LogRecord`.
+
+        Parameters
+        ----------
+        record : `logging.LogRecord`
+            The record from which to extract the relevant information.
+        """
+        # The properties that are one-to-one mapping.
+        simple = ("name", "levelno", "levelname", "filename", "pathname",
+                  "lineno", "funcName", "process", "processName")
+
+        record_dict = {k: getattr(record, k) for k in simple}
+
+        record_dict["message"] = record.getMessage()
+
+        # MDC
+        if hasattr(record, "MDC"):
+            # mypy does not understand hasattr.
+            record_dict["MDC"] = dict(record.MDC)  # type: ignore
+        else:
+            # Use an empty MDC to prevent format failures.
+            record_dict["MDC"] = defaultdict(str)
```

Comment:
Actually that's not quite the same because the original code makes a point of copying `record.MDC` into a new dict so that later changes to the MDC won't change the record after the fact. With copying I think it needs to be:

```
record_dict["MDC"] = dict(getattr(record, "MDC", {})
```
This does make the formatting a touch problematic if `LABEL` is not in MDC but I'm not sure what to do about that anyway since as written the record constructed from json won't be using defaultdict either.

Also, since this is a pydantic model it's a bit tricky to add keys at the top level so doing `record_dict(**record.MDC)` isn't really going to work.

Reply:
So if we need more dynamic model for this record type we cannot use pydantic? Like with any other persistent data we need to be careful with schema changes, at some point we can decide to add/remove attributes to this class and then we need to think about compatibility between new code and what is already stored in butler (or old code and data saved by new release). Maybe we need this class to be a simple dict with arbitrary set of keys (though it would just move compatibility issue to a different level)?

Reply:
Depends what you mean by dynamic. At the moment with pydantic I've set up the model so that all the LogRecord items are included (I'm keeping the exc_info text distinct from the message itself for now) and all the arbitrary stuff goes into the `MDC` attribute which can accept anything. That seems flexible enough to me. Maybe I should rename MDC to `extra`?  I've never ingested records into LogStash so I don't know how well it will deal with the `MDC` dict or whether it really wants `record.LABEL` to be used and the `MDC` hierarchy missing completely.  The `LogRecord` formatting of `MDC[LABEL]` is not the same code as a the `ButlerLogRecord` serialization so I can convert the `MDC` part of the dict to a defaultdict at that point when calling `ButlerLogRecord.format()`

Reply:
By dynamic I mean that someone could decide later that using `extra` keyword to logging methods is a great idea (in addition to MDC) and wants to save all those LogRecord attributes in the same JSON format as top-level keys. That set of attributes is potentially unbounded. Packing all of them into `MDC` or one `extra` attribute in this class may or may not work depending on how they are used, I guess we can only discover that later. We could probably reserve `extra` attribute for that future extension and hope that nesting all those extra attributes into one dictionary will work for everyone.